File: lib/core/api_client.dart
import 'dart:convert';
import 'dart:async';
import 'dart:io';

import 'package:http/http.dart' as http;

class ApiException implements Exception {
  const ApiException(this.statusCode, this.code, this.message);
  final int statusCode;
  final String code;
  final String message;

  @override
  String toString() => message;
}

typedef SessionReader =
    Future<({String? accessToken, String? refreshToken})> Function();
typedef SessionWriter =
    Future<void> Function(String accessToken, String refreshToken);
typedef SessionClearer = Future<void> Function();

class ApiClient {
  ApiClient({
    http.Client? httpClient,
    SessionReader? readSession,
    SessionWriter? writeSession,
    SessionClearer? clearSession,
    String? apiBaseUrl,
    Duration? requestTimeout,
  }) : _http = httpClient ?? http.Client(),
       _readSession = readSession,
       _writeSession = writeSession,
       _clearSession = clearSession,
       _baseUrl = (apiBaseUrl ?? baseUrl).replaceFirst(RegExp(r'/$'), ''),
       _requestTimeout = requestTimeout ?? const Duration(seconds: 20);

  static const baseUrl = String.fromEnvironment(
    'SEMITRACK_API_URL',
    defaultValue: 'http://10.0.2.2:4000',
  );
  final http.Client _http;
  final SessionReader? _readSession;
  final SessionWriter? _writeSession;
  final SessionClearer? _clearSession;
  final String _baseUrl;
  final Duration _requestTimeout;
  Future<bool>? _refreshInFlight;
  bool _closed = false;

  Future<Map<String, dynamic>> getJson(String path) => _request('GET', path);
  Future<Map<String, dynamic>> postJson(
    String path,
    Map<String, dynamic> body,
  ) => _request('POST', path, body: body);
  Future<Map<String, dynamic>> putJson(
    String path,
    Map<String, dynamic> body,
  ) => _request('PUT', path, body: body);
  Future<Map<String, dynamic>> patchJson(
    String path,
    Map<String, dynamic> body,
  ) => _request('PATCH', path, body: body);
  Future<void> delete(String path) async => _request('DELETE', path);

  Future<Map<String, dynamic>> _request(
    String method,
    String path, {
    Map<String, dynamic>? body,
    bool retryAfterRefresh = true,
  }) async {
    if (_closed) {
      throw const ApiException(
        0,
        'CLIENT_CLOSED',
        'This connection is no longer available.',
      );
    }
    if (!path.startsWith('/')) {
      throw const ApiException(
        0,
        'INVALID_PATH',
        'The requested API path is invalid.',
      );
    }
    final session = await _safeReadSession();
    final headers = <String, String>{'Accept': 'application/json'};
    if (body != null) headers['Content-Type'] = 'application/json';
    if (session.accessToken != null) {
      headers['Authorization'] = 'Bearer ${session.accessToken}';
    }
    final Uri uri;
    try {
      uri = Uri.parse('$_baseUrl$path');
    } on FormatException {
      throw const ApiException(
        0,
        'INVALID_API_URL',
        'SemiTrax is not configured with a valid server address.',
      );
    }
    final response = await _send(method, uri, headers, body);
    if (response.statusCode == 401 &&
        retryAfterRefresh &&
        session.refreshToken != null &&
        path != '/auth/refresh') {
      final refreshed = await _refresh();
      if (refreshed) {
        return _request(method, path, body: body, retryAfterRefresh: false);
      }
    }
    if (response.statusCode == 204 || response.body.isEmpty) return const {};
    final Object? decoded;
    try {
      decoded = jsonDecode(response.body);
    } on FormatException {
      throw ApiException(
        response.statusCode,
        'INVALID_RESPONSE',
        response.statusCode >= 200 && response.statusCode < 300
            ? 'SemiTrax received an invalid server response. Please try again.'
            : 'The server could not complete this request.',
      );
    }
    final json = decoded is Map<String, dynamic>
        ? decoded
        : <String, dynamic>{'data': decoded};
    if (response.statusCode < 200 || response.statusCode >= 300) {
      final error = json['error'];
      if (error is Map<String, dynamic>) {
        throw ApiException(
          response.statusCode,
          error['code']?.toString() ?? 'API_ERROR',
          error['message']?.toString() ?? 'Request failed',
        );
      }
      throw ApiException(
        response.statusCode,
        'API_ERROR',
        error?.toString() ?? 'Request failed',
      );
    }
    return json;
  }

  Future<http.Response> _send(
    String method,
    Uri uri,
    Map<String, String> headers,
    Map<String, dynamic>? body,
  ) async {
    try {
      final encodedBody = body == null ? null : jsonEncode(body);
      return await switch (method) {
        'POST' =>
          _http
              .post(uri, headers: headers, body: encodedBody)
              .timeout(_requestTimeout),
        'PUT' =>
          _http
              .put(uri, headers: headers, body: encodedBody)
              .timeout(_requestTimeout),
        'PATCH' =>
          _http
              .patch(uri, headers: headers, body: encodedBody)
              .timeout(_requestTimeout),
        'DELETE' =>
          _http.delete(uri, headers: headers).timeout(_requestTimeout),
        _ => _http.get(uri, headers: headers).timeout(_requestTimeout),
      };
    } on TimeoutException {
      throw const ApiException(
        0,
        'REQUEST_TIMEOUT',
        'The SemiTrax server took too long to respond. Please try again.',
      );
    } on SocketException {
      throw const ApiException(
        0,
        'CONNECTION_FAILED',
        'Unable to reach SemiTrax. Check your connection and try again.',
      );
    } on http.ClientException {
      throw const ApiException(
        0,
        'CONNECTION_FAILED',
        'Unable to reach SemiTrax. Check your connection and try again.',
      );
    }
  }

  Future<({String? accessToken, String? refreshToken})>
  _safeReadSession() async {
    try {
      return await _readSession?.call() ??
          (accessToken: null, refreshToken: null);
    } catch (_) {
      // A damaged/locked secure-storage entry must not block public endpoints.
      return (accessToken: null, refreshToken: null);
    }
  }

  Future<bool> _refresh() {
    return _refreshInFlight ??= _performRefresh().whenComplete(
      () => _refreshInFlight = null,
    );
  }

  Future<bool> _performRefresh() async {
    final session = await _safeReadSession();
    if (session.refreshToken == null || _writeSession == null) return false;
    try {
      final response = await _http
          .post(
            Uri.parse('$_baseUrl/auth/refresh'),
            headers: {
              'Content-Type': 'application/json',
              'Accept': 'application/json',
            },
            body: jsonEncode({'refreshToken': session.refreshToken}),
          )
          .timeout(_requestTimeout);
      if (response.statusCode < 200 || response.statusCode >= 300) {
        await _clearSession?.call();
        return false;
      }
      final decoded = jsonDecode(response.body);
      if (decoded is! Map<String, dynamic>) return false;
      final accessToken = decoded['accessToken'];
      final refreshToken = decoded['refreshToken'];
      if (accessToken is! String ||
          accessToken.isEmpty ||
          refreshToken is! String ||
          refreshToken.isEmpty) {
        return false;
      }
      await _writeSession(accessToken, refreshToken);
      return true;
    } catch (_) {
      return false;
    }
  }

  void close() {
    if (_closed) return;
    _closed = true;
    _http.close();
  }
}

File: apps/mobile-react-native/src/services/api/ApiClient.ts
import { sanitizeRestrictionDiagnostic } from '../../features/routing/restrictionDiagnostic';
import { safeDriverError } from '../../errors/driverErrors';
import { tokensSchema } from '../../models/contracts';
import type { TokenVault } from '../storage/TokenVault';
export class ApiError extends Error {
  constructor(
    readonly code: string,
    message: string,
    readonly status = 0,
    readonly retryable = false,
    readonly validationFields: string[] = [],
    readonly restrictionDiagnostic: ReturnType<
      typeof sanitizeRestrictionDiagnostic
    > = null,
  ) {
    super(message);
  }
}
type Reply = { status: number; ok: boolean; text: string };
export class ApiClient {
  private refreshFlight: Promise<boolean> | null = null;
  private sessionGeneration = 0;
  onUnauthorized: () => void = () => {};
  constructor(
    private baseUrl: string,
    private vault: TokenVault,
    private transport: typeof fetch = fetch,
  ) {}
  invalidateSession() {
    this.sessionGeneration++;
    this.refreshFlight = null;
  }
  /** App-private file bytes; authorization follows the same session/refresh boundary. */
  async uploadDocument(
    path: string,
    uri: string,
    onProgress: (percent: number) => void,
    signal?: AbortSignal,
  ): Promise<void> {
    if (
      !/^\/documents\/[^/]+\/attachments\/[a-f0-9-]+\/bytes$/.test(path) ||
      !uri.startsWith('file://')
    )
      throw new ApiError('INVALID_PATH', 'Invalid document upload.');
    const generation = this.sessionGeneration;
    const send = async () => {
      const tokens = await this.vault.read();
      if (generation !== this.sessionGeneration || !tokens)
        throw new ApiError('SESSION_CHANGED', 'Sign in again.');
      return new Promise<number>((resolve, reject) => {
        const xhr = new XMLHttpRequest();
        const abort = () => xhr.abort();
        const finish = (action: () => void) => {
          signal?.removeEventListener('abort', abort);
          action();
        };
        xhr.open('PUT', this.baseUrl + path);
        xhr.timeout = 90000;
        xhr.setRequestHeader('Authorization', 'Bearer ' + tokens.accessToken);
        xhr.setRequestHeader('Content-Type', 'application/octet-stream');
        xhr.upload.onprogress = event => {
          if (generation === this.sessionGeneration && event.lengthComputable)
            onProgress(
              Math.min(100, Math.floor((event.loaded / event.total) * 100)),
            );
        };
        xhr.onload = () => finish(() => resolve(xhr.status));
        xhr.onerror = () =>
          finish(() =>
            reject(
              new ApiError(
                'NETWORK_UNAVAILABLE',
                'Upload failed. Retry when connected.',
              ),
            ),
          );
        xhr.ontimeout = () =>
          finish(() =>
            reject(
              new ApiError(
                'REQUEST_TIMEOUT',
                'Upload timed out. Retry when connected.',
              ),
            ),
          );
        xhr.onabort = () =>
          finish(() =>
            reject(new ApiError('REQUEST_CANCELLED', 'Upload cancelled.')),
          );
        signal?.addEventListener('abort', abort);
        if (signal?.aborted) {
          finish(() =>
            reject(new ApiError('REQUEST_CANCELLED', 'Upload cancelled.')),
          );
          return;
        }
        // React Native's native networking accepts URI bodies without base64 in JS.
        xhr.send({ uri } as unknown as Parameters<XMLHttpRequest['send']>[0]);
      });
    };
    let status = await send();
    if (generation !== this.sessionGeneration)
      throw new ApiError('SESSION_CHANGED', 'Session changed.');
    if (status === 401 && (await this.refresh())) status = await send();
    if (generation !== this.sessionGeneration)
      throw new ApiError('SESSION_CHANGED', 'Session changed.');
    if (status === 401) {
      this.invalidateSession();
      await this.vault.clear();
      this.onUnauthorized();
    }
    if (status < 200 || status >= 300)
      throw new ApiError(
        'DOCUMENT_UPLOAD_FAILED',
        'Upload could not be confirmed. Retry.',
        status,
      );
  }
  async request<T = unknown>(
    method: string,
    path: string,
    body?: unknown,
    signal?: AbortSignal,
    authenticated = true,
    onRouteResponse?: (status: number) => void,
  ): Promise<T> {
    if (!path.startsWith('/') || path.startsWith('//') || path.includes('\\')) {
      throw new ApiError('INVALID_PATH', 'Invalid API path.');
    }
    const observeRouteResponse = (status: number) => {
      if (method === 'POST' && path === '/routing/truck-route') {
        try {
          onRouteResponse?.(status);
        } catch {
          /* Observability cannot change requests. */
        }
      }
    };
    const generation = this.sessionGeneration;
    const tokens = authenticated ? await this.vault.read() : null;
    if (authenticated && generation !== this.sessionGeneration) {
      throw new ApiError('SESSION_CHANGED', 'Session changed.');
    }
    let response = await this.send(
      method,
      path,
      body,
      tokens?.accessToken,
      signal,
    );
    observeRouteResponse(response.status);
    if (authenticated && generation !== this.sessionGeneration) {
      throw new ApiError('SESSION_CHANGED', 'Session changed.');
    }
    if (authenticated && response.status === 401 && tokens?.refreshToken) {
      const latest = await this.vault.read();
      const refreshed =
        latest !== null && latest.accessToken !== tokens.accessToken
          ? true
          : await this.refresh();
      if (refreshed && generation === this.sessionGeneration) {
        const retryTokens = await this.vault.read();
        if (generation !== this.sessionGeneration || !retryTokens)
          throw new ApiError('SESSION_CHANGED', 'Session changed.');
        response = await this.send(
          method,
          path,
          body,
          retryTokens.accessToken,
          signal,
        );
        observeRouteResponse(response.status);
      }
    }
    if (authenticated && generation !== this.sessionGeneration) {
      throw new ApiError('SESSION_CHANGED', 'Session changed.');
    }
    if (response.status === 401 && authenticated) {
      this.invalidateSession();
      await this.vault.clear();
      this.onUnauthorized();
    }
    let data: unknown;
    try {
      data = response.text ? JSON.parse(response.text) : null;
    } catch {
      if (!response.ok) {
        const code =
          response.status === 401
            ? 'AUTH_EXPIRED'
            : response.status === 502
            ? 'PROVIDER_UNAVAILABLE'
            : 'REQUEST_FAILED';
        throw new ApiError(
          code,
          safeDriverError({ code, status: response.status }),
          response.status,
          response.status === 429 || response.status >= 500,
        );
      }
      throw new ApiError(
        'INVALID_RESPONSE',
        'Invalid server response.',
        response.status,
      );
    }
    if (!response.ok) {
      const error =
        data && typeof data === 'object' && 'error' in data
          ? (data as { error: unknown }).error
          : null;
      const detail =
        error && typeof error === 'object'
          ? (error as {
              code?: unknown;
              message?: unknown;
              retryable?: unknown;
              restrictionDiagnostic?: unknown;
              details?: { fieldErrors?: unknown };
            })
          : {};
      throw new ApiError(
        typeof detail.code === 'string' ? detail.code : 'REQUEST_FAILED',
        safeDriverError({ code: detail.code, status: response.status }),
        response.status,
        typeof detail.retryable === 'boolean'
          ? detail.retryable
          : response.status === 429 || response.status >= 500,
        detail.code === 'VALIDATION_ERROR' &&
        detail.details?.fieldErrors &&
        typeof detail.details.fieldErrors === 'object'
          ? Object.keys(detail.details.fieldErrors).slice(0, 50)
          : [],
        detail.code === 'TRIMBLE_RESTRICTION_WARNING'
          ? sanitizeRestrictionDiagnostic(detail.restrictionDiagnostic)
          : null,
      );
    }
    return data as T;
  }
  /** Revoke only the captured session; never refresh or read a newer login. */
  async revokeSession(tokens: { accessToken: string; refreshToken: string }) {
    const response = await this.send(
      'POST',
      '/auth/logout',
      { refreshToken: tokens.refreshToken },
      tokens.accessToken,
    );
    if (!response.ok)
      throw new ApiError(
        'LOGOUT_REVOKE_FAILED',
        'Remote sign-out could not be confirmed.',
        response.status,
      );
  }
  private async send(
    method: string,
    path: string,
    body?: unknown,
    access?: string,
    signal?: AbortSignal,
  ): Promise<Reply> {
    if (signal?.aborted)
      throw new ApiError('REQUEST_CANCELLED', 'The request was cancelled.');
    const controller = new AbortController();
    let rejectInterrupted!: (error: Error) => void;
    const interrupted = new Promise<never>((_resolve, reject) => {
      rejectInterrupted = reject;
    });
    const abort = () => {
      controller.abort();
      rejectInterrupted(
        new ApiError('REQUEST_CANCELLED', 'The request was cancelled.'),
      );
    };
    signal?.addEventListener('abort', abort);
    if (signal?.aborted) abort();
    const timer = setTimeout(() => {
      controller.abort();
      rejectInterrupted(
        new ApiError(
          'REQUEST_TIMEOUT',
          'SemiTraX took too long to respond.',
          0,
          true,
        ),
      );
    }, 20000);
    try {
      return await Promise.race([
        interrupted,
        (async () => {
          const response = await this.transport(this.baseUrl + path, {
            method,
            signal: controller.signal,
            headers: {
              Accept: 'application/json',
              ...(body !== undefined
                ? { 'Content-Type': 'application/json' }
                : {}),
              ...(access ? { Authorization: 'Bearer ' + access } : {}),
            },
            ...(body !== undefined ? { body: JSON.stringify(body) } : {}),
          });
          const text = await response.text();
          if (text.length > 20_000_000)
            throw new ApiError(
              'INVALID_RESPONSE',
              'SemiTraX returned an oversized response.',
              response.status,
            );
          return { status: response.status, ok: response.ok, text };
        })(),
      ]);
    } catch (error) {
      if (error instanceof ApiError) throw error;
      throw new ApiError(
        signal?.aborted
          ? 'REQUEST_CANCELLED'
          : controller.signal.aborted
          ? 'REQUEST_TIMEOUT'
          : 'NETWORK_UNAVAILABLE',
        'Unable to reach SemiTraX. Check your connection and retry.',
        0,
        !signal?.aborted,
      );
    } finally {
      clearTimeout(timer);
      signal?.removeEventListener('abort', abort);
    }
  }
  private refresh() {
    if (this.refreshFlight) {
      return this.refreshFlight;
    }
    const flight = this.performRefresh().finally(() => {
      if (this.refreshFlight === flight) {
        this.refreshFlight = null;
      }
    });
    this.refreshFlight = flight;
    return flight;
  }
  private async performRefresh(): Promise<boolean> {
    const generation = this.sessionGeneration;
    const tokens = await this.vault.read();
    if (!tokens) {
      return false;
    }
    const response = await this.send('POST', '/auth/refresh', {
      refreshToken: tokens.refreshToken,
    });
    if (generation !== this.sessionGeneration) {
      return false;
    }
    if (response.status === 401 || response.status === 400) {
      this.invalidateSession();
      await this.vault.clear();
      this.onUnauthorized();
      return false;
    }
    if (!response.ok) {
      throw new ApiError(
        'REFRESH_UNAVAILABLE',
        'Session refresh is temporarily unavailable. Retry when connected.',
        response.status,
        true,
      );
    }
    let next;
    try {
      next = tokensSchema.parse(JSON.parse(response.text));
    } catch {
      throw new ApiError(
        'INVALID_RESPONSE',
        'The session response could not be validated.',
        response.status,
      );
    }
    if (generation !== this.sessionGeneration) {
      return false;
    }
    await this.vault.write(next);
    return generation === this.sessionGeneration;
  }
}
